File: Haskell/Implementations/Prototype/SpaceShip.hs
{-# LANGUAGE ExistentialQuantification #-}

module SpaceShip
    ( SpaceShip(..)
    , Part(..)
    , mkSpaceShip
    , addPart
    ) where

import Display (Display(..))

data Part = forall p . Display p => Part p

instance Display Part where
    display (Part p) = display p

data SpaceShip = MkSpaceShip
    { parts :: [Part]
    , name  :: String
    }

instance Display SpaceShip where
    display (MkSpaceShip ps n) =
        "Space ship "
            <> n
            <> " with "
            <> show (length ps)
            <> " parts:\n"
            <> mconcat ((\p -> "  " <> p <> "\n") . display <$> ps)

mkSpaceShip :: String -> SpaceShip
mkSpaceShip = MkSpaceShip []

addPart :: Display part => part -> SpaceShip -> SpaceShip
addPart p ss = ss { parts = parts ss <> [Part p] }

File: Java/Implementations/Prototype/SpaceShip.java
import java.util.*;

public class SpaceShip {
	private List<Part> parts = new ArrayList<>();
	private String name;

	public SpaceShip(String name) {
		this.name = name;
	}

	public SpaceShip addPart(Part part) {
		this.parts.add(part.clone());
		return this;
	}

	public String display() {
		StringJoiner sj = new StringJoiner("\n");
		sj.add(
			"Space ship "
			+ this.name
			+ " with "
			+ this.parts.size()
			+ " parts:"
		);
		for (Part part : this.parts) {
			sj.add("  " + part.display());
		}
		return sj.toString();
	}
}
